practical-react-unit-testing-with-jest-mocks-and-spies/Exercise5/solution/src/Vehicle.test.ts

```diff
@@ -0,0 +1,16 @@
+import Vehicle from "./Vehicle";
+
+jest.mock('./Vehicle', () => jest.fn().mockImplementation(() => ({
+    getVehicle: jest.fn().mockImplementation(() => "Wrangler")
+})));
+
+
+describe('Vehicle component', () => {
+    it('uses global mock', () => {
+      const instance = new Vehicle('Renegade');
+      instance.getVehicle();
+
+      expect(instance.getVehicle).toHaveBeenCalled();
+      expect(instance.getVehicle()).toEqual("Wrangler")
```

Comment:
This test feels really contrived to me because it doesn’t matter what `Vehicle` has in it because our mock will be called.

Maybe a good exercise would be something where a class will call another class, so we want to replace the second class with our own mock so we can focus on the first class. Here’s my attempt to make it a little more realistic:

Let’s say we have this `Vehicle` class:

```ts
import VehicleDatabase from './VehicleDatabase';

export default class Vehicle {
  database: VehicleDatabase;
  vehicleType: string;

  constructor(type: string) {
    this.database = new VehicleDatabase();
    this.vehicleType = type;
  }

  getVehicleDetails() {
    return this.database.getVehicleDetails(this.vehicleType);
  }
}
```

It depends on this `VehicleDatabase` class:

```ts
export default class VehicleDatabase {
    getVehicleDetails(type: string) {
        // In reality, this would call a database or an API endpoint.
        console.info(`Called getVehicleDetails with type: ${type}`);
        throw new Error("You should not call this directly without mocking!");
    }
}
```

In the tests, we could have something like this:

```ts
import Vehicle from "./Vehicle";

/*

Exercise: mock the VehicleDatabase’s implementation here.

*/
jest.mock('./VehicleDatabase', () => jest.fn().mockImplementation(() => ({
  getVehicleDetails: jest.fn().mockReturnValue("Wrangler details")
})));

describe('Vehicle class', () => {
  it('fetches vehicle details from mocked database', () => {
    const vehicle = new Vehicle('Wrangler');
    const details = vehicle.getVehicleDetails();

    expect(details).toEqual("Wrangler details");
  });
});
```